src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -162,28 +159,44 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 }
 
                 if (NetEventSource.IsEnabled) Trace($"{_request.Content}");
-
                 Debug.Assert(_requestBodyCancellationSource != null);
 
-                // Create a linked cancellation token source so that we can cancel the request in the event of receiving RST_STREAM
-                // and similiar situations where we need to cancel the request body (see Cancel method).
-                _requestBodyCancellationToken = cancellationToken.CanBeCanceled ?
-                    CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _requestBodyCancellationSource.Token).Token :
-                    _requestBodyCancellationSource.Token;
-
+                // Cancel the request body sending if cancellation is requested on the supplied cancellation token.
+                // Normally we might create a linked token, but once cancellation is requested, we can't recover anyway,
+                // so it's fine to cancel the source representing the whole request body, and doing so allows us to avoid
+                // creating another CTS instance and the associated nodes inside of it.  With this, cancellation will be
+                // requested on _requestBodyCancellationSource when we need to cancel the request stream for any reason,
+                // such as receiving an RST_STREAM or when the passed in token has cancellation requested. However, to
+                // avoid unnecessarily registering with the cancellation token unless we have to, we wait to do so until
+                // either we know we need to do a Expect: 100-continue send or until we know that the copying of our
+                // content completed asynchronously.
+                CancellationTokenRegistration linkedRegistration = default;
                 try
                 {
                     bool sendRequestContent = true;
                     if (_expect100ContinueWaiter != null)
                     {
-                        sendRequestContent = await WaitFor100ContinueAsync(_requestBodyCancellationToken).ConfigureAwait(false);
+                        linkedRegistration = RegisterRequestBodyCancellation(cancellationToken);
+                        sendRequestContent = await WaitFor100ContinueAsync(_requestBodyCancellationSource.Token).ConfigureAwait(false);
                     }
 
                     if (sendRequestContent)
                     {
-                        using (Http2WriteStream writeStream = new Http2WriteStream(this))
+                        using var writeStream = new Http2WriteStream(this);
+
+                        ValueTask vt = _request.Content.InternalCopyToAsync(writeStream, context: null, _requestBodyCancellationSource.Token);
+                        if (vt.IsCompleted)
```

Comment:
Is there a reason why this is ``IsCompleted`` instead of ``IsCompletedSuccessfully``? What happens if the task has an error?

Reply:
If there's an exception it'll propagate out of the GetResult() call just as it would have out of the await. There's no reason to limit this to only the success case.